File: src/typegate/src/engine/typecheck/inline_validators/file.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import type { FileNode } from "../../../typegraph/types.ts";
import { check } from "./common.ts";
import {
  type ConstraintSpec,
  generateConstraintValidatorsFor,
} from "./constraints.ts";

const fileConstraints = [
  {
    name: "minSize",
    test: (ref, c) => `${ref}.size < ${c}`,
    description: "minimum size",
  },
  {
    name: "maxSize",
    test: (ref, c) => `${ref}.size > ${c}`,
    description: "maximum size",
  },
  {
    name: "mimeTypes",
    test: (ref, c) => `${JSON.stringify(c)}.includes(${ref}.type)`,
    description: "mime type",
  },
] as const satisfies ConstraintSpec[];

export function generateFileValidator(
  typeNode: FileNode,
  varName: string,
  path: string,
): string[] {
  return [
    check(`${varName} instanceof File`, `"At ${path}: expected File"`),
    ...generateConstraintValidatorsFor(fileConstraints, typeNode),
  ];
}

File: src/typegraph/graph/src/types/file.rs
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::Arc;

use super::{TypeBase, TypeNode};

#[derive(Debug)]
pub struct FileType {
    pub base: TypeBase,
    pub min_size: Option<u32>,
    pub max_size: Option<u32>,
    pub mime_types: Option<Vec<String>>,
}

impl TypeNode for Arc<FileType> {
    fn base(&self) -> &TypeBase {
        &self.base
    }

    fn tag(&self) -> &'static str {
        "file"
    }
}

impl FileType {
    pub fn is_plain(&self) -> bool {
        matches!(
            (self.min_size, self.max_size, self.mime_types.as_ref()),
            (None, None, None)
        )
    }
}
